src/pulsar_producer.erl

```diff
@@ -49,233 +92,640 @@ callback_mode() -> [state_functions].
 
-send(Pid, Message) ->
-    gen_statem:cast(Pid, {send, Message}).
-
-send_sync(Pid, Message) ->
-    send_sync(Pid, Message, 5000).
+-spec send(gen_statem:server_ref(), [pulsar:message()]) -> ok.
+send(Pid, Messages) ->
+    From = undefined,
+    erlang:send(Pid, ?SEND_REQ(From, Messages)),
+    ok.
 
-send_sync(Pid, Message, Timeout) ->
-    gen_statem:call(Pid, {send, Message}, Timeout).
+-spec send_sync(gen_statem:server_ref(), [pulsar:message()]) ->
+          {ok, send_receipt()}
+        | {error, producer_connecting
+                | producer_disconnected
+                | term()}.
+send_sync(Pid, Messages) ->
+    send_sync(Pid, Messages, 5_000).
+
+-spec send_sync(gen_statem:server_ref(), [pulsar:message()], timeout()) ->
+          {ok, send_receipt()}
+        | {error, producer_connecting
+                | producer_disconnected
+                | term()}.
+send_sync(Pid, Messages, Timeout) ->
+    Caller = self(),
+    MRef = erlang:monitor(process, Pid),
+    %% Mimicking gen_statem's From, so the reply can be sent with
+    %% `gen_statem:reply/2'
+    From = {Caller, MRef},
+    erlang:send(Pid, ?SEND_REQ(From, Messages)),
+    receive
+        {MRef, Response} ->
+            erlang:demonitor(MRef, [flush]),
+            Response;
+        {'DOWN', MRef, process, Pid, Reason} ->
+            error({producer_down, Reason})
+    after
+        Timeout ->
+            erlang:demonitor(MRef, [flush]),
+            receive
+                {MRef, Response} ->
+                    Response
+            after
+                0 ->
+                    error(timeout)
+            end
+    end.
 
 %%--------------------------------------------------------------------
-%% gen_server callback
+%% gen_statem callback
 %%--------------------------------------------------------------------
-init([PartitionTopic, Server, ProxyToBrokerUrl, ProducerOpts]) ->
+
+-spec init({string(), string(), string() | undefined, config()}) ->
+          gen_statem:init_result(statem(), state()).
+init({PartitionTopic, Server, ProxyToBrokerUrl, ProducerOpts0}) ->
+    process_flag(trap_exit, true),
     {Transport, BrokerServer} = pulsar_utils:parse_url(Server),
-    State = #state{
-        partitiontopic = PartitionTopic,
-        producer_id = maps:get(producer_id, ProducerOpts),
-        producer_name = maps:get(producer_name, ProducerOpts, pulsar_producer),
-        callback = maps:get(callback, ProducerOpts, undefined),
-        batch_size = maps:get(batch_size, ProducerOpts, 0),
-        broker_server = BrokerServer,
-        opts = pulsar_utils:maybe_enable_ssl_opts(Transport, ProducerOpts)
+    ProducerID = maps:get(producer_id, ProducerOpts0),
+    ReplayqCfg0 =
+        case maps:get(replayq_dir, ProducerOpts0, false) of
+            false ->
+                #{mem_only => true};
+            BaseDir ->
+                Dir = filename:join([BaseDir, PartitionTopic]),
```

Comment:
is it allowed to have special chars in pulsar topic ? e.g. `/root` this will make filename:join result in path `/root`

Reply:
![2022-09-21_10-47](https://user-images.githubusercontent.com/16166434/191521246-34007527-5498-4d6d-9695-ccb77c1ab50a.png)

`replayq` at least seems able to create the directory correctly when there are non-ASCII chars in the topic.  Pulsar also seems to cope with it.  I produced and consumed a message to that `🍺topic`. 

It doesn't display very well in the `pulsar-admin` output, though.

```
I have no name!@664933b1c40a:/pulsar$ bin/pulsar-admin topics list public/default
persistent://public/default/t_pulsar_drop_expired_batch
persistent://public/default/?topic
persistent://public/default/t_pulsar_replayq
persistent://public/default/t_pulsar_replayq_producer_restart
```

Reply:
I'm more concerned about the leading `/` of a pulsar topic name.
this will cause any filename:join to take no effect.
e.g. `filename:join(["whatever/prefix", "/whaterver/abs/path"])` yields `/whatever/abs/path`.

Reply:
we should maybe sanitise the path by encoding special chars to base64 for example.

Reply:
Aha, I see.  Maybe we can normalize the topic when constructing the path by simply removing any leading slashes?

If we just remove special characters or substitute them for placeholders (like `_`), there might be some clashes between similar topic names?

Reply:
> we should maybe sanitise the path by encoding special chars to base64 for example.

This seems like a better idea.  This way, we don't lose any information and avoid clashes.

Reply:
:heavy_check_mark: